syntax = "proto3";

// ----------------------------------------------------
// Service 1: Kernel (Client Gateway)
// User (Client SDK) talks to this 
// ----------------------------------------------------
service FSMasterService {
    rpc Mount(MountRequest) returns (StatusResponse);
    rpc Open(OpenRequest) returns (OpenResponse);
    rpc Close(CloseRequest) returns (StatusResponse);
    
    rpc Read(ReadRequest) returns (ReadResponse);
    rpc Write(WriteRequest) returns (StatusResponse);
    rpc DeleteFile(DeleteFileRequest) returns (StatusResponse);
    
    rpc Mkdir(MkdirRequest) returns (StatusResponse);
    rpc Rmdir(RmdirRequest) returns (StatusResponse);
    rpc Ls(LsRequest) returns (LsResponse);
}
// 1)
message MountRequest { string user_id = 1; }
message OpenRequest  { string user_id = 1; string path = 2; string mode = 3; }
message OpenResponse { int32 fd = 1; string error = 2; }
message CloseRequest { string user_id = 1; string path = 2; }
// 2)
message ReadRequest {
    string user_id = 1;
    int32 fd = 2;
    int32 count = 3;
}
message ReadResponse {
    bytes data = 1;
    int32 bytes_read = 2;
}
message WriteRequest {
    string user_id = 1;
    int32 fd = 2;
    bytes data = 3;
}
message DeleteFileRequest {
    string user_id = 1;
    string path = 2;
}
// 3)
message MkdirRequest { string user_id = 1; string path = 2; }
message RmdirRequest { string user_id = 1; string path = 2; }
message LsRequest    { string user_id = 1; string path = 2; }
message LsResponse   { repeated string files = 1; }

// ----------------------------------------------------
// Service 2: FS Server (Raw Data)
// Kernel talks to this to read/write bytes
// ----------------------------------------------------
service FSServerService {
    rpc ReadBlock(BlockRequest) returns (ReadBlockResponse);
    rpc WriteBlock(BlockRequest) returns (StatusResponse);
}
message BlockRequest  { uint64 block_uuid = 1; }
message ReadBlockResponse { bool success = 1; bytes data = 2; }

// common
message StatusResponse {
    bool success = 1;
    string error = 2;
}